From 0053d9128efa313e4d5ce0cd761336ef5e845b48 Mon Sep 17 00:00:00 2001
From: Ahmad Masri <amasri@codeaurora.org>
Date: Mon, 15 Feb 2021 09:50:12 +0200
Subject: [PATCH] wil6210: Drop plaintext frames on secure network

On secure network, only eap frames are valid unencrypted frames and
allowed to be indicated to the local network, any other unencrypted
frame should be dropped immediately.

Change-Id: Id9d97a4c0984f7bf2d7d6941c4c61e87bc2354cd
Signed-off-by: Ahmad Masri <amasri@codeaurora.org>
---
 drivers/net/wireless/ath/wil6210/txrx.c | 11 +++++++++++
 1 file changed, 11 insertions(+)

diff --git a/drivers/net/wireless/ath/wil6210/txrx.c b/drivers/net/wireless/ath/wil6210/txrx.c
index 58f0da56bdab4..f9d63ffcd54c3 100644
--- a/drivers/net/wireless/ath/wil6210/txrx.c
+++ b/drivers/net/wireless/ath/wil6210/txrx.c
@@ -872,6 +872,17 @@ void wil_netif_rx_any(struct sk_buff *skb, struct net_device *ndev)
 
 	skb_orphan(skb);
 
+	/* pass only EAPOL packets as plaintext */
+	if (wil->privacy && !security &&
+	    wil_skb_get_protocol(skb) != htons(ETH_P_PAE)) {
+		wil_dbg_txrx(wil,"Rx drop plaintext frame with %d bytes in secure network\n",
+			     skb->len);
+		dev_kfree_skb(skb);
+		ndev->stats.rx_dropped++;
+		stats->rx_dropped++;
+		return;
+	}
+
 	if (security && (wil_rx_crypto_check(wil, skb) != 0)) {
 		rc = GRO_DROP;
 		dev_kfree_skb(skb);
-- 
2.28.0

